docker/pulsar/scripts/install-glibc-compat.sh

```diff
@@ -18,13 +18,18 @@
 # under the License.
 #
 
-set -x
+set -e -x
 
-# TODO: remove these lines once grpcio doesn't need to compile from source on ARM64 platform
-ARCH=$(uname -m | sed -r 's/aarch64/arm64/g' |  awk '!/arm64/{$0="amd64"}1')
-if [ "${ARCH}" == "arm64" ]; then
-  apt update
-  apt -y install build-essential python3-dev
+ARCH=$(uname -m)
+if [ "$ARCH" == "x86_64" ]
+then
+  cd /tmp
+  echo 'https://storage.sev.monster/alpine/edge/testing' | tee -a /etc/apk/repositories
+  wget https://storage.sev.monster/alpine/edge/testing/x86_64/sevmonster-keys-1-r0.apk
+  apk add --no-cache --allow-untrusted ./sevmonster-keys-1-r0.apk
+  apk update
+  apk add gcompat libuuid
+  rm /lib/ld-linux-x86-64.so.2
+  apk add --no-cache --force-overwrite glibc glibc-bin
+  rm *.apk
```

Comment:
Why is this needed? Where is the homepage of this repository? Isn't `gcompat`, `libgcc` and `libstdc++`/`libstdc++6` sufficient alone?
It looks like the sevmonster repository doesn't contain arm64 support.

Is it a security risk to trust the sevmonster repository?

If we need true glibc, there's also the possibility to use [Bellsoft's JDK base images](https://bell-sw.com/libericajdk-containers/#docker-hub-images) which support JDK on glibc on Alpine.

This is [liberica-openjdk-alpine](https://hub.docker.com/r/bellsoft/liberica-openjdk-alpine)'s Dockerfile: https://github.com/bell-sw/Liberica/blob/master/docker/repos/liberica-openjdk-alpine/21/Dockerfile

Liberica JDK is also used in [Apache Nifi's Dockerfile](https://github.com/apache/nifi/blob/main/nifi-docker/dockerhub/Dockerfile#L20).

Reply:
Does Alpine support `libcirce-checksum.so`?

On my Linux, box this is what I see as the dependencies:
```
❯ ldd libcirce-checksum.so
	linux-vdso.so.1 (0x00007fff45ba0000)
	libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007b4a61c00000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007b4a61800000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007b4a61f11000)
	/lib64/ld-linux-x86-64.so.2 (0x00007b4a6201f000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007b4a61ef1000)
```

Reply:
Looks like linking is fine for libcirce-checksum.so
```
❯ docker run --rm -it merlimat/pulsar:3.3.0-SNAPSHOT-f2a91a1 bash -c 'cd /tmp;unzip /pulsar/lib/org.apache.bookkeeper-circe-checksum-4.16.4.jar lib/*;ldd lib/libcirce-checksum.so'
Archive:  /pulsar/lib/org.apache.bookkeeper-circe-checksum-4.16.4.jar
   creating: lib/
  inflating: lib/libcirce-checksum.so
	/lib/ld-musl-x86_64.so.1 (0x7a6468661000)
	libstdc++.so.6 => /usr/lib/libstdc++.so.6 (0x7a64683c3000)
	libc.so.6 => /lib/ld-musl-x86_64.so.1 (0x7a6468661000)
	libgcc_s.so.1 => /usr/lib/libgcc_s.so.1 (0x7a646839f000)
```

Reply:
On Linux x86_64:
```
❯ docker run --rm -it merlimat/pulsar:3.3.0-SNAPSHOT-f2a91a1 bash -c 'apk list|grep glibc'
glibc-2.38-r0 x86_64 {glibc} (LGPL) [installed]
glibc-bin-2.38-r0 x86_64 {glibc} (LGPL) [installed]
glibc-dev-2.38-r0 x86_64 {glibc} (LGPL)
glibc-i18n-2.38-r0 x86_64 {glibc} (LGPL)
```

On Mac Apple silicon (arm64):
```
❯ docker run --rm -it merlimat/pulsar:3.3.0-SNAPSHOT-f2a91a1 bash -c 'apk list|grep glibc'
WARNING: opening from cache https://dl-cdn.alpinelinux.org/alpine/v3.19/main: No such file or directory
WARNING: opening from cache https://dl-cdn.alpinelinux.org/alpine/v3.19/community: No such file or directory
WARNING: opening from cache https://apk.corretto.aws: No such file or directory
```

It seems that there's a difference in the x86_64 and arm64 images.

Reply:
Who is behind https://storage.sev.monster/alpine/edge/testing ? I couldn't find any information by googling. The whois information for the sev.monster domain is hidden.

Reply:
It looks like the sev.monster might be publishing similar packages as https://github.com/sgerrand/alpine-pkg-glibc / https://github.com/sgerrand/docker-glibc-builder . However, I'd prefer using the [bellsoft/liberica-openjdk-alpine:21](https://hub.docker.com/r/bellsoft/liberica-openjdk-alpine) base image since it has a reliable configuration for glibc for both amd64 & arm64.

Reply:
The most problem I had were with with Kinesis producer lib (for the Pulsar-IO kinesis connector). It uses boost and a whole bunch of dependencies on Glibc. 

> It looks like the sev.monster might be publishing similar packages as https://github.com/sgerrand/alpine-pkg-glibc / https://github.com/sgerrand/docker-glibc-builder . 

I was just trying to get a quick solution to see that Kinesis producer was working. I saw they had last prebuilt latest version, since the other ones on  https://github.com/sgerrand/alpine-pkg-glibc are not working on Alpine 3.19. 

I definitely agree we shouldn't depend on these binary repo. 

I would still prefer to use Corretto JVM linked with musl. Mainly so that: 
 1. we can always control the base OS version, without depending on this OpenJDK provider to have a release
 2. the track record of Corretto in the last years
 
In the end, I would prefer to just build the Glibc as a pre-step, so we'd be having both x86_64 and arm binaries and no dependencies.

Reply:
Just as a nod to the usage of Liberica..

Spring Boot uses Liberica as the default choice for JDK and images for its end-end testing.

Reply:
> I would still prefer to use Corretto JVM linked with musl. Mainly so that:
> 
> 1. we can always control the base OS version, without depending on this OpenJDK provider to have a release

I'd assume that Liberica has more resources for testing the base image's compatibility making it a "certified" option. Isn't that relevant for us?

Why would we need to control the base OS version? If we really need to do that, that could be done by building the base image in a similar way as Liberica is doing it. 

I'm assuming that we have a use case where we need to use glibc because of different reasons such as the Kinesis producer example. 

[glibc's malloc has it's warts in running Java](https://bugs.openjdk.org/browse/JDK-8193521) and even Bellsoft provides alternative malloc implementations in their Alpaquita Linux base images (https://docs.bell-sw.com/alpaquita-linux/latest/how-to/malloc/).  Bellsoft's Alpaquita is similar to Alpine, but not compatible. It is optimized for running Java in containers (https://bell-sw.com/alpaquita-linux/).

[Bellsoft has some perf graphs](https://bell-sw.com/assets/files/WP_Alpaquita_Linux.pdf#page=8) showing that standard musl is slower in certain microbenchmarks compared to glibc and their's "musl perf" which is an optimized version of musl. I'm not sure how valid that would be to Pulsar / Bookkeeper use cases. Perhaps there isn't a significant difference at all.

Staying on glibc seems like a safe bet. I don't see what benefits switching to musl would bring. The part that is not optimal for Java in glibc is the malloc implementation (and there are multiple options to replace that, jemalloc as one of the most popular ones). In the bellsoft/liberica-openjdk-alpine:21 base image, glibc takes about 7MB of uncompressed space in the image so the additional size of glibc isn't a problem for our use cases.

What does musl give us that we don't have with glibc?

> 2. the track record of Corretto in the last years

What makes Corretto significantly better than Liberica OpenJDK? Is there something outstanding in Corretto that isn't in Liberica OpenJDK (or Eclipse Temurin)?

> In the end, I would prefer to just build the Glibc as a pre-step, so we'd be having both x86_64 and arm binaries and no dependencies.

What do you mean as a pre-step? Isn't it going to be a problem if a single process would be using both glibc and musl? Let's say that a dynamically linked JNI library would require glibc for some reason and the Java process would be using musl. Does that really work and make sense?